cli/cluster/cmd/common.go

```diff
@@ -184,65 +184,75 @@ func connectEtcd(uriOpts UriOpts, connOpts connectOpts) (*clientv3.Client, error
+	if errTarantool == nil {
+		return tarantoolFunc(conn)
+	}
+
+	etcdcli, errEtcd := connectEtcd(opts, connOpts)
+	if errEtcd == nil {
+		return etcdFunc(etcdcli)
+	}
+
+	return fmt.Errorf("failed to establish a connection to tarantool or etcd: %w, %w",
+		errTarantool, errEtcd)
+}
+
+// createPublisherAndCollector creates a new data publisher and collector based on UriOpts.
 func createPublisherAndCollector(
 	publishers libcluster.DataPublisherFactory,
 	collectors libcluster.CollectorFactory,
 	connOpts connectOpts,
 	opts UriOpts) (libcluster.DataPublisher, libcluster.Collector, func(), error) {
 	prefix, key, timeout := opts.Prefix, opts.Key, opts.Timeout
 
-	conn, errTarantool := connectTarantool(opts, connOpts)
-	if errTarantool == nil {
-		var (
-			publisher libcluster.DataPublisher
-			collector libcluster.Collector
-			err       error
-		)
+	var (
+		publisher libcluster.DataPublisher
+		collector libcluster.Collector
+		err       error
+		closeFunc func()
+	)
+
+	tarantoolFunc := func(conn tarantool.Connector) error {
 		if publishers != nil {
 			publisher, err = publishers.NewTarantool(conn, prefix, key, timeout)
 			if err != nil {
-				return nil, nil, nil,
-					fmt.Errorf("failed to create tarantool config storage publisher: %w", err)
+				return fmt.Errorf("failed to create tarantool config storage publisher: %w", err)
 			}
 		}
-
 		if collectors != nil {
 			collector, err = collectors.NewTarantool(conn, prefix, key, timeout)
 			if err != nil {
-				return nil, nil, nil,
-					fmt.Errorf("failed to create tarantool config storage collector: %w", err)
+				return fmt.Errorf("failed to create tarantool config storage collector: %w", err)
 			}
 		}
-		return publisher, collector, func() { conn.Close() }, nil
+		closeFunc = func() { conn.Close() }
+		return nil
 	}
 
-	etcdcli, errEtcd := connectEtcd(opts, connOpts)
-	if errEtcd == nil {
-		var (
-			publisher libcluster.DataPublisher
-			collector libcluster.Collector
-			err       error
-		)
+	etcdFunc := func(client *clientv3.Client) error {
 		if publishers != nil {
-			publisher, err = publishers.NewEtcd(etcdcli, prefix, key, timeout)
+			publisher, err = publishers.NewEtcd(client, prefix, key, timeout)
 			if err != nil {
-				return nil, nil, nil,
-					fmt.Errorf("failed to create etcd publisher: %w", err)
+				return fmt.Errorf("failed to create etcd publisher: %w", err)
 			}
 		}
-
 		if collectors != nil {
-			collector, err = collectors.NewEtcd(etcdcli, prefix, key, timeout)
+			collector, err = collectors.NewEtcd(client, prefix, key, timeout)
 			if err != nil {
-				return nil, nil, nil,
-					fmt.Errorf("failed to create etcd collector: %w", err)
+				return fmt.Errorf("failed to create etcd collector: %w", err)
 			}
 		}
-		return publisher, collector, func() { etcdcli.Close() }, nil
+		closeFunc = func() { client.Close() }
+		return nil
+	}
+
+	if err := selectStorage(connOpts, opts, tarantoolFunc, etcdFunc); err != nil {
+		return nil, nil, nil, err
```

Comment:
An opened connection is not closed in the case (and in some other error cases above).

Reply:
Thank you! `Close()` if error is added.